docs: Improve the abortable broadcast documentation (#3260)
dfinity/ic: rs/p2p/consensus_manager/README.adoc

= Abortable Broadcast 
:toc:  

== Background 

[quote, Internet Computer for Geeks]
To guarantee liveness, the Internet Computer Protocol assumes a form of partial synchrony, which (roughly stated) says that the network will
be periodically synchronous for short intervals of time. In such intervals of synchrony, all undelivered messages will be delivered in less than time δ, for some fixed bound δ. 
The bound δ does not have to be known in advance (the protocol is initialized with a reasonable bound, but will dynamically adapt and increase this bound if it is too small).

> Regardless of whether we are assuming an asynchronous or a partially synchronous network, we assume that every message sent from one honest replica to another will eventually be delivered.

Reviewer: ```suggestion
Regardless of whether we are assuming an asynchronous or a partially synchronous network, we assume that every message sent from one honest replica to another will eventually be delivered and received by all other honest replicas.
```

Author: I believe, "delivered" implies exactly the same as "received by all other honest replicas"

> Regardless of whether we are assuming an asynchronous or a partially synchronous network, we assume that every message sent from one honest replica to another will eventually be delivered.

Reviewer 2: > Regardless of whether we are assuming an asynchronous or a partially synchronous network, 
This describes the underlying network model the proofs are based on.

> we assume that every message sent from one honest replica to another will eventually be delivered.
This is confusing. Consensus proofs rely on all messages being delivered. P2P has to provide this guarantee. It's not something we can assume the underlying network will provide, but something we actively work for with retransmissions so we actually can guarantee it.

Author: This is literally copy-pasted from [ICP for geeks](https://internetcomputer.org/whitepaper.pdf)

See the sentence before "5.2 Protocol overview"

Reviewer 2: Section 5 is about the consensus layer and its assumptions. If we want to keep this quote, we need to make it clear here, that it describes what the consensus layer expects from P2P, or said differently, that P2P provides this (P2P is also part of the IC protocol suite...)

As we know from the work on abortable broadcast, we don't need EVERY message to be eventually delivered. Therefore I don't think that this quote here is really helpful as is.  

Moreover, since safety is at least as crucial as liveness, I'd mention both or none of them. To make it more clear that we don't talk about the assumptions P2P relies on, but the consensus layer above, what about the following phrasing of the paragraph?

"Most properties of the protocols of the Consensus layer hold under an asynchronous network model, with no a priori bound on the delay of messages sent between replicas. To guarantee liveness, the Internet Computer Consensus Protocol assumes a form of partial synchrony,...
...
Regardless of whether an asynchronous or a partially synchronous network is assumed, the Consensus layer protocols rely on the fact that messages sent from one honest replica to another will eventually be delivered unless the transmission is explicitly aborted."

Author: Thanks for the observation. You are correct!

== Overview  

This library provides an implementation of the https://arxiv.org/abs/2410.22080[Abortable Broadcast] primitive.

> This mechanism ensures that any message sent from one honest party to all other honest parties will eventually be delivered, provided the message is still active (i.e., it has not been aborted).

Reviewer: Nit:
```suggestion
This mechanism ensures that any message sent from one honest party to all other honest parties will eventually be delivered, provided the transmission of the message is still active (i.e., it has not been aborted).
```

Author: I would keep as is, because "transmission" is a bit of an implementation detail

=== Key Features

> * Delivery Guarantee: Messages sent between honest parties are reliably delivered as long as they remain active.

Reviewer: ```suggestion
* Delivery Guarantee: Messages sent between honest parties are reliably delivered as long as the message transmission remains active.
```

Author: ditto

* Dynamic Adaptability: The implementation dynamically adjusts to changing network conditions, ensuring reliable message delivery without requiring fixed, pre-defined timing bounds.
* Fault Tolerance: The implementation is robust against Byzantine, malicious, or otherwise faulty peers. It avoids slowing down the producer of data and prevents unbounded message buffering, even in the presence of adversarial behavior.
* Finite Active Messages Assumption: The design achieves efficiency and simplicity by leveraging the assumption that there is always a finite number of active messages in the system. This assumption ensures predictable resource usage and prevents attacks that might exploit excessive message production.

> * Simplicity: The interface is designed to be intuitive, exposing a channel-like API for publishing messages to all peers in the subnet.

Reviewer: ```suggestion
* Simplicity: The interface is designed to be intuitive, exposing a channel API for publishing messages to all peers in the subnet.
```

> This design makes the abortable broadcast primitive well-suited for environments with partial synchrony, where the protocol must handle a mix of honest and potentially adversarial behavior while remaining reliable, efficient, and easy to use.

Reviewer 2: The design also works fine for asynchrony, doesn't it?

Author: Yes.

Reviewer 2: ```suggestion
This design makes the abortable broadcast primitive well-suited for asynchronous or partially synchronous environments, where the protocol must handle a mix of honest and potentially adversarial behavior while remaining reliable, efficient, and easy to use.
```
